src/fsharp/FSharp.Core/tasks.fs

```diff
@@ -0,0 +1,424 @@
+        // new struct type.  Meth1 and Meth2 are used to implement the methods on the interface implemented by the struct type.
+        let __stateMachineStruct<'Template, 'Meth1, 'Meth2, 'Result> (_meth1: 'Meth1) (_meth2: 'Meth2) (_after: unit -> 'Result): 'Result = failwith "__stateMachineStruct should always be removed from compiled code"
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __newEntryPoint() : int = failwith "__newEntryPoint should always be removed from compiled code"
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __machine<'T> : 'T = failwith "__newEntryPoint should always be removed from compiled code"
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __machineAddr<'T> : byref<'T> = (# "ldnull throw" : byref<'T> #)
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __entryPoint (_n: int) : unit = failwith "__entryPoint should always be removed from compiled code" 
+
+        [<MethodImpl(MethodImplOptions.NoInlining)>]
+        let __return<'T> (_v: 'T) : 'T = failwith "__return should always be removed from compiled code"
+#endif
+
+#if !BUILDING_WITH_LKG && !BUILD_FROM_SOURCE
+namespace Microsoft.FSharp.Control
+
+open System
+open System.Runtime.CompilerServices
+open System.Threading.Tasks
+open Microsoft.FSharp.Core
+open Microsoft.FSharp.Core.Printf
+open Microsoft.FSharp.Core.CompilerServices
+open Microsoft.FSharp.Core.CompilerServices.CodeGenHelpers
+open Microsoft.FSharp.Core.LanguagePrimitives.IntrinsicOperators
+open Microsoft.FSharp.Control
+open Microsoft.FSharp.Collections
+
+/// Represents the state of a computation: either awaiting something with a
+/// continuation, or completed with a return value.
+//
+// Uses a struct-around-single-reference to allow future changes in representation (the representation is
+// not revealed in the signature)
+[<Struct; NoComparison; NoEquality>]
+type TaskStep<'T, 'TOverall>(completed: bool) = 
+    member x.IsCompleted = completed
+
+[<Struct; NoComparison; NoEquality>]
+type TaskStateMachineTemplate<'T> =
+
+    [<DefaultValue(false)>]
+    val mutable Result : 'T
+
+    [<DefaultValue>]
+    val mutable ResumptionPoint : int
+
+    [<DefaultValue>]
+    val mutable MethodBuilder : AsyncTaskMethodBuilder<'T>
+
+    interface IAsyncStateMachine with 
+        member this.MoveNext() = failwith "template"
+        member this.SetStateMachine(_machine) = failwith "template"
+
+module Helpers =
+    let inline Start (sm: byref<TaskStateMachineTemplate<'T>>) = 
+        //Console.WriteLine("[{0}] start", sm.GetHashCode())
+        sm.MethodBuilder <- AsyncTaskMethodBuilder<'T>.Create()
+        sm.MethodBuilder.Start(&sm)
+        //Console.WriteLine("[{0}] unwrap", sm.GetHashCode())
+        sm.MethodBuilder.Task
+
+    let inline Await (sm: byref<TaskStateMachineTemplate<'T>>) (awaiter, pc) =
+        sm.ResumptionPoint <- pc
+        let mutable awaiter = awaiter
```

Comment:
Should be able to remove this if awaiter is passed byref, see later comments

Reply:
I think it's not possible - the method expects a `byref` not `inref`

Reply:
A byref argument isn't an inref though?

Reply:
Done now, thanks!